src/runtime-rs/crates/hypervisor/src/qemu/qmp.rs

```diff
@@ -65,6 +72,227 @@ impl Qmp {
+                .qom_get_slot(parent_bus)
+                .context("qom get parent slot failed")?;
+
+            // Prepend the slot
+            slots.insert(0, parent_slot);
+            parent_path = parent_bus.to_string();
+        }
+
+        // It's safe to directly unwrap as slots is some.
+        Ok(PciPath::new(slots).unwrap())
+    }
+
+    pub fn execute_vfio_device_add(
+        &mut self,
+        bus: &str,
+        id: &str,
+        driver: &str,
+        bdf: &str,
+        romfile: &str,
+    ) -> Result<()> {
+        // if transport.isVirtioCCW(nil) {
+        //     driver = string(VfioCCW)
+        // } else {
+        //     driver = string(VfioPCI)
+        // }
+        info!(
+            sl!(),
+            "execute_vfio_device_add bus: {:?}, id: {:?}, driver: {:?}, host: {:?}",
+            bus,
+            id,
+            driver,
+            bdf
+        );
+        let mut vfio_args = Dictionary::new();
+
+        let normalize = |bdf: &str| -> String {
+            if !bdf.starts_with("0000") {
+                format!("0000:{}", bdf)
+            } else {
+                bdf.to_string()
+            }
+        };
+
+        if !bdf.is_empty() {
+            vfio_args.insert("host".to_owned(), normalize(bdf).into());
+        }
+
+        if !romfile.is_empty() {
+            vfio_args.insert("romfile".to_owned(), romfile.into());
+        }
+
+        // "bus: \"rp0\", id: \"vfio_device_0\", driver: \"vfio-pci\", host: \"0000:c1:00.0\"
+        let try_status = self.qmp.execute(&qmp::query_status {}).unwrap();
+        info!(sl!(), "qmp status: {:?}", try_status);
+
+        let vfio_device_add = qmp::device_add {
+            bus: if !bus.is_empty() {
+                Some(bus.to_string())
+            } else {
+                None
+            },
+            id: Some(id.to_string()),
+            driver: driver.to_string(),
+            arguments: vfio_args,
+        };
+
+        self.qmp
+            .execute(&vfio_device_add)
+            .map_err(|e| anyhow!(format!("{:?}", e)))?;
+
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn execute_pci_vfio_device_add(
+        &mut self,
+        bus: &str,
+        id: &str,
+        driver: &str,
+        addr: &str,
+        romfile: &str,
+        bdf: &str,
+    ) -> Result<()> {
+        let mut vfio_args = Dictionary::new();
+        vfio_args.insert("host".to_owned(), bdf.into());
+        vfio_args.insert("romfile".to_owned(), romfile.into());
+        vfio_args.insert("addr".to_owned(), addr.into());
+
+        self.qmp.execute(&qmp::device_add {
+            bus: if !bus.is_empty() {
+                Some(bus.to_string())
+            } else {
+                None
+            },
+            id: Some(id.to_string()),
+            driver: driver.to_string(),
+            arguments: vfio_args,
+        })?;
+
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn execute_pci_vfio_mediated_device_add(
```

Comment:
`execute_vfio_device_add()`, `execute_pci_vfio_device_add()` and `execute_pci_vfio_mediated_device_add()` strike me as very similar, doing pretty much the same with some slight variations and quite a lot of logic duplication.  Is there any reason not to refactor them into a single function?

Reply:
Thx Pavel. Yes agree with your point, I will update this PR days later. As  I have a higher priority task to do now.

Reply:
I would offer taking this over to reduce your load, however I don't think I have a good way to test this functionality... :-(